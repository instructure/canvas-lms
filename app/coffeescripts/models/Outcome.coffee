#
# Copyright (C) 2012 Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.
#

define [
  'i18n!outcomes'
  'underscore'
  'Backbone'
  'compiled/models/grade_summary/CalculationMethodContent'
], (I18n, _, Backbone, CalculationMethodContent) ->

  class Outcome extends Backbone.Model
    defaults:
      mastery_points: 3
      points_possible: 5
      ratings: [
        description: I18n.t("criteria.exceeds_expectations", "Exceeds Expectations")
        points: 5
      ,
        description: I18n.t("criteria.meets_expectations", "Meets Expectations")
        points: 3
      ,
        description: I18n.t("criteria.does_not_meet_expectations", "Does Not Meet Expectations")
        points: 0
      ]

    defaultCalculiationInt: -> {
      n_mastery: 5
      decaying_average: 65
    }[@get('calculation_method')]

    initialize: ->
      @setDefaultCalcSettings() unless @get('calculation_method')
      @on 'change:calculation_method', (model, changedTo) =>
        model.set calculation_int: @defaultCalculiationInt()
      super

    setDefaultCalcSettings: ->
      @set {
        calculation_method: 'decaying_average'
        calculation_int: '65'
      }

    calculationMethodContent: ->
      new CalculationMethodContent(@)

    calculationMethods: ->
      @calculationMethodContent().toJSON()

    name: ->
      @get 'title'

    isNative: ->
      @outcomeLink && (@get('context_id') == @outcomeLink.context_id && @get('context_type') == @outcomeLink.context_type)

    # The api returns abbreviated data by default
    # which in most cases means there's no description.
    # Run fetch() to get all the data.
    isAbbreviated: ->
      !@has('description')

    # overriding to work with both outcome and outcome link responses
    parse: (resp) ->
      if resp.outcome # it's an outcome link
        @outcomeLink = resp
        @outcomeGroup = resp.outcome_group
        resp.outcome
      else
        resp

    present: ->
      _.extend({}, @toJSON(), @calculationMethodContent().present())

    setUrlTo: (action) ->
      @url =
        switch action
          when 'add'    then @outcomeGroup.outcomes_url
          when 'edit'   then @get 'url'
          when 'delete' then @outcomeLink.url
