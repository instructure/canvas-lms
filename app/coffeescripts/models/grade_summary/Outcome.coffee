define [
  'i18n!outcomes'
  'underscore'
  'Backbone'
  'compiled/models/grade_summary/CalculationMethodContent'
], (I18n, _, {Model, Collection}, CalculationMethodContent) ->

  class Outcome extends Model
    defaults:
      calculation_method: "highest"

    initialize: ->
      super
      @set 'friendly_name', @get('display_name') || @get('title')
      @set 'hover_name', (@get('title') if @get('display_name'))

    status: ->
      if @scoreDefined()
        score = @get('score')
        mastery = @get('mastery_points')
        if score >= mastery + (mastery / 2)
          'exceeds'
        else if score >= mastery
          'mastery'
        else if score >= mastery / 2
          'near'
        else
          'remedial'
      else
        'undefined'

    statusTooltip: ->
      {
        'undefined': I18n.t('Unstarted')
        'remedial': I18n.t('Well Below Mastery')
        'near': I18n.t('Near Mastery')
        'mastery': I18n.t('Meets Mastery')
        'exceeds': I18n.t('Exceeds Mastery')
      }[@status()]

    roundedScore: ->
      score = @get('score')
      if _.isNumber(score)
        Math.round(score * 100.0) / 100.0
      else
        null

    scoreDefined: ->
      _.isNumber(@get('score'))

    percentProgress: ->
      if @scoreDefined()
        @get('score')/@get('points_possible') * 100
      else
        0

    masteryPercent: ->
      @get('mastery_points')/@get('points_possible') * 100

    present: ->
      _.extend({}, @toJSON(), new CalculationMethodContent(@).present())

    toJSON: ->
      _.extend super,
        status: @status()
        statusTooltip: @statusTooltip()
        roundedScore: @roundedScore()
        scoreDefined: @scoreDefined()
        percentProgress: @percentProgress()
        masteryPercent: @masteryPercent()
