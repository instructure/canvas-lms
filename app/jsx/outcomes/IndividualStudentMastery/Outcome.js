/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import PropTypes from 'prop-types'
import I18n from 'i18n!outcomes'
import View from '@instructure/ui-layout/lib/components/View'
import Flex, { FlexItem } from '@instructure/ui-layout/lib/components/Flex'
import ToggleGroup from '@instructure/ui-toggle-details/lib/components/ToggleGroup'
import List, { ListItem } from '@instructure/ui-elements/lib/components/List'
import Pill from '@instructure/ui-elements/lib/components/Pill'
import Text from '@instructure/ui-elements/lib/components/Text'
import IconOutcomes from '@instructure/ui-icons/lib/Line/IconOutcomes'
import natcompare from 'compiled/util/natcompare'
import AssignmentResult from './AssignmentResult'
import * as shapes from './shapes'

export default class Outcome extends React.Component {
  static propTypes = {
    outcome: shapes.outcomeShape.isRequired,
    expanded: PropTypes.bool.isRequired,
    onExpansionChange: PropTypes.func.isRequired,
    outcomeProficiency: shapes.outcomeProficiencyShape
  }

  static defaultProps = {
    outcomeProficiency: null
  }

  handleToggle = (_event, expanded) => {
    this.props.onExpansionChange('outcome', this.props.outcome.expansionId, expanded)
  }

  renderHeader () {
    const { outcome } = this.props
    const { mastered, results, title } = outcome
    const numAlignments = results.length

    return (
      <Flex direction="row" justifyItems="space-between">
        <FlexItem>
          <Flex direction="column">
            <FlexItem>
              <Text size="medium">
                <Flex>
                  <FlexItem><IconOutcomes /></FlexItem>
                  <FlexItem padding="0 x-small">{ title }</FlexItem>
                </Flex>
              </Text>
            </FlexItem>
            <FlexItem><Text size="small">{ I18n.t({
              zero: 'No alignments',
              one: '%{count} alignment',
              other: '%{count} alignments'
            }, { count: I18n.n(numAlignments) }) }</Text></FlexItem>
          </Flex>
        </FlexItem>
        <FlexItem>
        {
          mastered ? <Pill text={I18n.t('Mastered')} variant="success" /> : <Pill text={I18n.t('Not mastered')} />
        }
        </FlexItem>
      </Flex>
    )
  }

  renderDetails () {
    const { outcome, outcomeProficiency } = this.props
    const { results } = outcome
    return (
      <List variant="unstyled" delimiter="dashed">
      {
          results.sort(natcompare.byKey('submitted_or_assessed_at')).reverse().map((result) => (
            <ListItem key={result.id}>
              <AssignmentResult result={result} outcome={outcome} outcomeProficiency={outcomeProficiency} />
            </ListItem>
        ))
      }
      </List>
    )
  }

  renderEmpty () {
    return (
      <View as="div" padding="small">
        <Text>{ I18n.t('No alignments are available for this outcome.') }</Text>
      </View>
    )
  }

  render () {
    const { outcome, expanded } = this.props
    const { results, title } = outcome
    const hasResults = results.length > 0
    return (
      <ToggleGroup
        summary={this.renderHeader()}
        toggleLabel={I18n.t('Toggle alignment details for %{title}', { title })}
        expanded={expanded}
        onToggle={this.handleToggle}
        border={false}
      >
        { hasResults ? this.renderDetails() : this.renderEmpty() }
      </ToggleGroup>
    )
  }
}
