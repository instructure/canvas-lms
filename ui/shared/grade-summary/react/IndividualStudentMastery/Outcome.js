/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import _ from 'lodash'
import PropTypes from 'prop-types'
import I18n from 'i18n!IndividualStudentMasteryOutcome'
import {Flex} from '@instructure/ui-flex'
import {View} from '@instructure/ui-view'
import {ToggleGroup} from '@instructure/ui-toggle-details'
import {Text} from '@instructure/ui-text'
import {Pill} from '@instructure/ui-pill'
import {List} from '@instructure/ui-list'
import natcompare from '@canvas/util/natcompare'
import AssignmentResult from './AssignmentResult'
import UnassessedAssignment from './UnassessedAssignment'
import OutcomePopover from './OutcomePopover'
import {ScreenReaderContent, PresentationContent} from '@instructure/ui-a11y-content'
import TruncateWithTooltip from '../TruncateWithTooltip'
import WithBreakpoints, {breakpointsShape} from 'with-breakpoints'
import * as shapes from './shapes'

class Outcome extends React.Component {
  static propTypes = {
    outcome: shapes.outcomeShape.isRequired,
    expanded: PropTypes.bool.isRequired,
    onExpansionChange: PropTypes.func.isRequired,
    outcomeProficiency: shapes.outcomeProficiencyShape,
    breakpoints: breakpointsShape
  }

  static defaultProps = {
    outcomeProficiency: null,
    breakpoints: {}
  }

  handleToggle = (_event, expanded) => {
    this.props.onExpansionChange('outcome', this.props.outcome.expansionId, expanded)
  }

  renderScoreAndPill() {
    const {outcome} = this.props
    const {mastered, score, points_possible, results} = outcome
    const pillAttributes = {text: I18n.t('Not mastered')}
    if (mastered) {
      Object.assign(pillAttributes, {text: I18n.t('Mastered'), variant: 'success'})
    }

    return (
      <Flex direction="row" justifyItems="start" padding="0 0 0 x-small">
        {_.isNumber(score) && !_.every(results, ['hide_points', true]) && (
          <Flex.Item padding="0 x-small 0 0">
            <span>
              <PresentationContent>
                <Text size="medium">
                  {score}/{points_possible}
                </Text>
              </PresentationContent>
              <ScreenReaderContent>
                {I18n.t('%{score} out of %{points_possible} points', {
                  score,
                  points_possible
                })}
              </ScreenReaderContent>
            </span>
          </Flex.Item>
        )}
        <Flex.Item>
          <Pill {...pillAttributes} />
        </Flex.Item>
      </Flex>
    )
  }

  renderHeader() {
    const {outcome, outcomeProficiency, breakpoints} = this.props
    const {assignments, display_name, title} = outcome
    const numAlignments = assignments.length
    const verticalLayout = !breakpoints.tablet

    return (
      <Flex
        direction={verticalLayout ? 'column' : 'row'}
        justifyItems={verticalLayout ? null : 'space-between'}
        alignItems={verticalLayout ? 'stretch' : null}
        data-selenium="outcome"
      >
        <Flex.Item grow as="div">
          <Text size="medium">
            <Flex>
              <Flex.Item>
                <OutcomePopover outcome={outcome} outcomeProficiency={outcomeProficiency} />
              </Flex.Item>
              <Flex.Item shrink>
                <TruncateWithTooltip>{display_name || title}</TruncateWithTooltip>
              </Flex.Item>
            </Flex>
          </Text>
          {verticalLayout && this.renderScoreAndPill()}
          <View as="div" padding="0 0 0 x-small">
            <Text size="small">
              {I18n.t(
                {
                  zero: 'No alignments',
                  one: '%{count} alignment',
                  other: '%{count} alignments'
                },
                {count: I18n.n(numAlignments)}
              )}
            </Text>
          </View>
        </Flex.Item>
        {!verticalLayout && <Flex.Item>{this.renderScoreAndPill()}</Flex.Item>}
      </Flex>
    )
  }

  renderDetails() {
    const {outcome, outcomeProficiency} = this.props
    const {assignments, results} = outcome
    const assignmentsWithResults = _.filter(results, r =>
      r.assignment.id.startsWith('assignment_')
    ).map(r => r.assignment.id.split('_')[1])
    const assessmentsWithResults = _.filter(results, r =>
      r.assignment.id.startsWith('live_assessments/assessment_')
    ).map(r => r.assignment.id.split('_')[2])
    const unassessed = _.filter(
      assignments,
      a =>
        (a.assignment_id && !_.includes(assignmentsWithResults, a.assignment_id.toString())) ||
        (a.assessment_id && !_.includes(assessmentsWithResults, a.assessment_id.toString()))
    )
    return (
      <List variant="unstyled" delimiter="dashed">
        {results
          .sort(natcompare.byKey('submitted_or_assessed_at'))
          .reverse()
          .map(result => (
            <List.Item key={result.id}>
              <AssignmentResult
                result={result}
                outcome={outcome}
                outcomeProficiency={outcomeProficiency}
              />
            </List.Item>
          ))}
        {unassessed.map(assignment => (
          <List.Item key={`a${assignment.assignment_id}`}>
            <UnassessedAssignment assignment={assignment} />
          </List.Item>
        ))}
      </List>
    )
  }

  renderEmpty() {
    return (
      <View as="div" padding="small">
        <Text>{I18n.t('No alignments are available for this outcome.')}</Text>
      </View>
    )
  }

  render() {
    const {outcome, expanded} = this.props
    const {assignments, title} = outcome
    const hasAlignments = assignments.length > 0
    return (
      <ToggleGroup
        summary={this.renderHeader()}
        toggleLabel={I18n.t('Toggle alignment details for %{title}', {title})}
        expanded={expanded}
        onToggle={this.handleToggle}
        border={false}
      >
        {hasAlignments ? this.renderDetails() : this.renderEmpty()}
      </ToggleGroup>
    )
  }
}

export default WithBreakpoints(Outcome)
