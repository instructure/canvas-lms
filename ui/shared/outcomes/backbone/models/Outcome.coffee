#
# Copyright (C) 2012 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.
#

import I18n from 'i18n!modelsOutcome'
import _ from 'underscore'
import Backbone from '@canvas/backbone'
import CalculationMethodContent from '@canvas/grade-summary/backbone/models/CalculationMethodContent'

export default class Outcome extends Backbone.Model
  defaults:
    mastery_points: 3
    points_possible: 5
    ratings: [
      description: I18n.t("criteria.exceeds_expectations", "Exceeds Expectations")
      points: 5
    ,
      description: I18n.t("criteria.meets_expectations", "Meets Expectations")
      points: 3
    ,
      description: I18n.t("criteria.does_not_meet_expectations", "Does Not Meet Expectations")
      points: 0
    ]

  setMasteryScales: ->
    ratings = ENV.MASTERY_SCALE.outcome_proficiency.ratings
    @set {
      ratings: ratings
      mastery_points: _.find(ratings, (r) => r.mastery).points
      points_possible: Math.max(_.map(ratings, (r) -> r.points)...)
    }

  defaultCalculationInt: -> {
    n_mastery: 5
    decaying_average: 65
  }[@get('calculation_method')]

  initialize: ->
    @setDefaultCalcSettings() unless @get('calculation_method')
    @setMasteryScales() if ENV.ACCOUNT_LEVEL_MASTERY_SCALES && ENV.MASTERY_SCALE?.outcome_proficiency
    @on 'change:calculation_method', (model, changedTo) =>
      model.set calculation_int: @defaultCalculationInt()
    super

  setDefaultCalcSettings: ->
    @set {
      calculation_method: 'decaying_average'
      calculation_int: '65'
    }

  calculationMethodContent: ->
    new CalculationMethodContent(@)

  calculationMethods: ->
    @calculationMethodContent().toJSON()

  name: ->
    @get 'title'

  canManage: ->
    @get('can_edit') || @canManageInContext()

  canManageInContext: ->
    ENV.ROOT_OUTCOME_GROUP?.context_type == "Course" && ENV.PERMISSIONS?.manage_outcomes && ENV.current_user_roles?.includes('admin')

  isNative: ->
    @outcomeLink && (@get('context_id') == @outcomeLink.context_id && @get('context_type') == @outcomeLink.context_type)

  # The api returns abbreviated data by default
  # which in most cases means there's no description.
  # Run fetch() to get all the data.
  isAbbreviated: ->
    !@has('description')

  # overriding to work with both outcome and outcome link responses
  parse: (resp) ->
    if resp.outcome # it's an outcome link
      @outcomeLink = resp
      @outcomeGroup = resp.outcome_group
      resp.outcome
    else
      resp

  present: ->
    _.extend({}, @toJSON(), @calculationMethodContent().present())

  setUrlTo: (action) ->
    @url =
      switch action
        when 'add'    then @outcomeGroup.outcomes_url
        when 'edit'   then @get 'url'
        when 'delete' then @outcomeLink.url

  save: (data, saveOpts) ->
    if ENV.ACCOUNT_LEVEL_MASTERY_SCALES
      @unset 'mastery_points'
      @unset 'points_possible'
      @unset 'ratings'
      @unset 'calculation_method'
      @unset 'calculation_int'
    super
