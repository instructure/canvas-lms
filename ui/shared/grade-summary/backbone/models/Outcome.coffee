#
# Copyright (C) 2014 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

import I18n from 'i18n!grade_summaryOutcome'
import _ from 'underscore'
import {Model, Collection} from '@canvas/backbone'
import Outcome from '@canvas/outcomes/backbone/models/Outcome.coffee'
import tz from '@canvas/timezone'


GradeSummary = {}
export default class GradeSummary.Outcome extends Outcome
  initialize: ->
    super
    @set 'friendly_name', @get('display_name') || @get('title')
    @set 'hover_name', (@get('title') if @get('display_name'))
    @set 'scaled_score', (@scaledScore())

  parse: (response) ->
    super _.extend(response, {
      submitted_or_assessed_at: tz.parse(response.submitted_or_assessed_at),
      question_bank_result: response.links?.alignment?.includes("assessment_question_bank")
    })

  status: ->
    if @scoreDefined()
      score = @score()
      mastery = @get('mastery_points')
      if score >= mastery + (mastery / 2)
        'exceeds'
      else if score >= mastery
        'mastery'
      else if score >= mastery / 2
        'near'
      else
        'remedial'
    else
      'undefined'

  statusTooltip: ->
    {
      'undefined': I18n.t('Unstarted')
      'remedial': I18n.t('Well Below Mastery')
      'near': I18n.t('Near Mastery')
      'mastery': I18n.t('Meets Mastery')
      'exceeds': I18n.t('Exceeds Mastery')
    }[@status()]

  roundedScore: ->
    if @scoreDefined()
      Math.round(@score() * 100.0) / 100.0
    else
      null

  scoreDefined: ->
    _.isNumber(@get('score'))

  scaledScore: ->
    is_aggregate_score = @get('question_bank_result')
    return unless @scoreDefined() && is_aggregate_score
    if @get('points_possible') > 0
      @get('percent') * @get('points_possible')
    else
      @get('percent') * @get('mastery_points')

  score: ->
    @get('scaled_score') || @get('score')

  percentProgress: ->
    return 0 unless @scoreDefined()
    if @get('percent')
      @get('percent') * 100
    else
      @score()/@get('points_possible') * 100

  masteryPercent: ->
    @get('mastery_points')/@get('points_possible') * 100

  toJSON: ->
    _.extend super,
      status: @status()
      statusTooltip: @statusTooltip()
      roundedScore: @roundedScore()
      scoreDefined: @scoreDefined()
      percentProgress: @percentProgress()
      masteryPercent: @masteryPercent()
