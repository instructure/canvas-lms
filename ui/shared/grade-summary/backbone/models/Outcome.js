/*
 * Copyright (C) 2023 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/* eslint-disable no-void */

import {extend} from '@canvas/backbone/utils'
import {useScope as useI18nScope} from '@canvas/i18n'

import _ from 'underscore'

import Outcome from '@canvas/outcomes/backbone/models/Outcome'

import tz from '@canvas/timezone'

const I18n = useI18nScope('grade_summaryOutcome')

extend(GradeSummaryOutcome, Outcome)

function GradeSummaryOutcome() {
  return GradeSummaryOutcome.__super__.constructor.apply(this, arguments)
}

GradeSummaryOutcome.prototype.initialize = function () {
  GradeSummaryOutcome.__super__.initialize.apply(this, arguments)
  this.set('friendly_name', this.get('display_name') || this.get('title'))
  this.set('hover_name', this.get('display_name') ? this.get('title') : void 0)
  return this.set('scaled_score', this.scaledScore())
}

GradeSummaryOutcome.prototype.parse = function (response) {
  let ref, ref1
  return GradeSummaryOutcome.__super__.parse.call(
    this,
    _.extend(response, {
      submitted_or_assessed_at: tz.parse(response.submitted_or_assessed_at),
      question_bank_result:
        (ref = response.links) != null
          ? (ref1 = ref.alignment) != null
            ? ref1.includes('assessment_question_bank')
            : void 0
          : void 0,
    })
  )
}

GradeSummaryOutcome.prototype.status = function () {
  let mastery, score
  if (this.scoreDefined()) {
    score = this.score()
    mastery = this.get('mastery_points')
    if (score >= mastery + mastery / 2) {
      return 'exceeds'
    } else if (score >= mastery) {
      return 'mastery'
    } else if (score >= mastery / 2) {
      return 'near'
    } else {
      return 'remedial'
    }
  } else {
    return 'undefined'
  }
}

GradeSummaryOutcome.prototype.statusTooltip = function () {
  return {
    undefined: I18n.t('Unstarted'),
    remedial: I18n.t('Well Below Mastery'),
    near: I18n.t('Near Mastery'),
    mastery: I18n.t('Meets Mastery'),
    exceeds: I18n.t('Exceeds Mastery'),
  }[this.status()]
}

GradeSummaryOutcome.prototype.roundedScore = function () {
  if (this.scoreDefined()) {
    return Math.round(this.score() * 100.0) / 100.0
  } else {
    return null
  }
}

GradeSummaryOutcome.prototype.scoreDefined = function () {
  return _.isNumber(this.get('score'))
}

GradeSummaryOutcome.prototype.scaledScore = function () {
  const is_aggregate_score = this.get('question_bank_result')
  if (!(this.scoreDefined() && is_aggregate_score)) {
    return
  }
  if (this.get('points_possible') > 0) {
    return this.get('percent') * this.get('points_possible')
  } else {
    return this.get('percent') * this.get('mastery_points')
  }
}

GradeSummaryOutcome.prototype.score = function () {
  return this.get('scaled_score') || this.get('score')
}

GradeSummaryOutcome.prototype.percentProgress = function () {
  if (!this.scoreDefined()) {
    return 0
  }
  if (this.get('percent')) {
    return this.get('percent') * 100
  } else {
    return (this.score() / this.get('points_possible')) * 100
  }
}

GradeSummaryOutcome.prototype.masteryPercent = function () {
  return (this.get('mastery_points') / this.get('points_possible')) * 100
}

GradeSummaryOutcome.prototype.toJSON = function () {
  return _.extend(GradeSummaryOutcome.__super__.toJSON.apply(this, arguments), {
    status: this.status(),
    statusTooltip: this.statusTooltip(),
    roundedScore: this.roundedScore(),
    scoreDefined: this.scoreDefined(),
    percentProgress: this.percentProgress(),
    masteryPercent: this.masteryPercent(),
  })
}

export default GradeSummaryOutcome
