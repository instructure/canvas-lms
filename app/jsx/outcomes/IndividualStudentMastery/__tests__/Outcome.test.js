/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import { render, shallow } from 'enzyme'
import Outcome from '../Outcome'

const result = (id = 1, date = new Date()) => ({
  id,
  percent: 0.1,
  assignment: {
    id: 1,
    html_url: 'http://foo',
    name: 'My alignment',
    submission_types: '',
    score: 0
  },
  submitted_or_assessed_at: date.toISOString()
})

const defaultProps = (props = {}) => (
  Object.assign({
    outcome: {
      id: 1,
      expansionId: 100,
      mastered: false,
      mastery_points: 3,
      points_possible: 5,
      ratings: [
        { description: 'My first rating' },
        { description: 'My second rating' }
      ],
      results: [
        {
          id: 1,
          score: 1,
          percent: 0.1,
          assignment: {
            id: 1,
            html_url: 'http://foo',
            name: 'My assignment',
            submission_types: 'online_quiz',
            score: 0
          }
        }
      ],
      title: 'My outcome'
    },
    expanded: false,
    onExpansionChange: () => {}
  }, props)
)

it('renders the Outcome component', () => {
  const wrapper = shallow(<Outcome {...defaultProps()}/>)
  expect(wrapper.debug()).toMatchSnapshot()
})

it('renders correctly expanded', () => {
  const wrapper = shallow(<Outcome {...defaultProps()} expanded />)
  expect(wrapper.debug()).toMatchSnapshot()
})

it('renders correctly expanded with no results', () => {
  const props = defaultProps()
  props.outcome.results = []
  const wrapper = shallow(<Outcome {...props} expanded />)
  expect(wrapper.debug()).toMatchSnapshot()
})

describe('header', () => {
  it('includes the outcome name', () => {
    const wrapper = shallow(<Outcome {...defaultProps()}/>)
    const header = wrapper.find('ToggleGroup')
    const summary = render(header.prop('summary'))
    expect(summary.text()).toMatch('My outcome')
  })

  it('includes mastery when mastered', () => {
    const props = defaultProps()
    props.outcome.mastered = true
    const wrapper = shallow(<Outcome {...props} />)
    const header = wrapper.find('ToggleGroup')
    const summary = render(header.prop('summary'))
    expect(summary.text()).toMatch('Mastered')
  })

  it('includes non-mastery when not mastered', () => {
    const wrapper = shallow(<Outcome {...defaultProps()} />)
    const header = wrapper.find('ToggleGroup')
    const summary = render(header.prop('summary'))
    expect(summary.text()).toMatch('Not mastered')
  })

  it('shows correct number of alignments', () => {
    const wrapper = shallow(<Outcome {...defaultProps()} />)
    const header = wrapper.find('ToggleGroup')
    const summary = render(header.prop('summary'))
    expect(summary.text()).toMatch('1 alignment')
  })
})

it('includes the individual results', () => {
  const wrapper = shallow(<Outcome {...defaultProps()} />)
  expect(wrapper.find('AssignmentResult')).toHaveLength(1)
})

it('renders the results by most recent', () => {
  const props = defaultProps()
  const now = new Date()
  const minuteAgo = new Date(now - 60000)
  const hourAgo = new Date(now - 3600000)
  const yearishAgo = new Date(now - 3600000 * 24 * 360)
  props.outcome.results = [
    result(1, hourAgo),
    result(2, now),
    result(3, minuteAgo),
    result(4, yearishAgo)
  ]

  const wrapper = shallow(<Outcome {...props} />)
  const results = wrapper.find('AssignmentResult')
  expect(results).toHaveLength(4)
  expect(results.get(0).props.result.id).toEqual(2) // now
  expect(results.get(1).props.result.id).toEqual(3) // minuteAgo
  expect(results.get(2).props.result.id).toEqual(1) // hourAgo
  expect(results.get(3).props.result.id).toEqual(4) // yearishAgo
})

describe('handleToggle()', () => {
  it('calls onExpansionChange with the correct data', () => {
    const props = defaultProps()
    props.onExpansionChange = jest.fn()

    const wrapper = shallow(<Outcome {...props} />)
    wrapper.instance().handleToggle(null, true)
    expect(props.onExpansionChange).toBeCalledWith('outcome', 100, true)
  })
})
