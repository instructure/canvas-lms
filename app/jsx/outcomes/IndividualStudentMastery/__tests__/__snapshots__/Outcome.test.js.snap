// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly expanded 1`] = `
"<ToggleGroup summary={{...}} toggleLabel=\\"Toggle alignment details for My outcome\\" expanded={true} onToggle={[Function]} border={false} size=\\"medium\\" icon={[Function]} iconExpanded={[Function]} defaultExpanded={false} transition={true} as=\\"span\\" elementRef={[Function]}>
  <List variant=\\"unstyled\\" delimiter=\\"dashed\\" as=\\"ul\\" margin=\\"none\\" size=\\"medium\\">
    <ListItem>
      <AssignmentResult result={{...}} outcome={{...}} outcomeProficiency={{...}} />
    </ListItem>
  </List>
</ToggleGroup>"
`;

exports[`renders correctly expanded with no results 1`] = `
"<ToggleGroup summary={{...}} toggleLabel=\\"Toggle alignment details for My outcome\\" expanded={true} onToggle={[Function]} border={false} size=\\"medium\\" icon={[Function]} iconExpanded={[Function]} defaultExpanded={false} transition={true} as=\\"span\\" elementRef={[Function]}>
  <View as=\\"div\\" padding=\\"small\\" display=\\"auto\\">
    <Text as=\\"span\\" size=\\"medium\\" letterSpacing=\\"normal\\">
      No alignments are available for this outcome.
    </Text>
  </View>
</ToggleGroup>"
`;

exports[`renders the Outcome component 1`] = `
"<ToggleGroup summary={{...}} toggleLabel=\\"Toggle alignment details for My outcome\\" expanded={false} onToggle={[Function]} border={false} size=\\"medium\\" icon={[Function]} iconExpanded={[Function]} defaultExpanded={false} transition={true} as=\\"span\\" elementRef={[Function]}>
  <List variant=\\"unstyled\\" delimiter=\\"dashed\\" as=\\"ul\\" margin=\\"none\\" size=\\"medium\\">
    <ListItem>
      <AssignmentResult result={{...}} outcome={{...}} outcomeProficiency={{...}} />
    </ListItem>
  </List>
</ToggleGroup>"
`;
